Instance generators

jdumouchelle opened the pull request:
## Pull request checklist
<!---
Thank you for contributing to Ecole!

⚠️ To ensure a smooth review, please make sure to complete the following checklist. ⚠️

Except for very small fixes, please open an issue to discuss the changes and fill
the number bellow.
Refer to the README for help on running the tests, checks, formatters, and documentation.
-->
- [x] I have open an issue to discuss the proposed changes. Fix #82 
- [x] I have modified/added tests to cover the new changes/features.
- [x] I have modified/added the documenation to cover the new changes/features.
- [x] I have ran the tests, checks, and code formatters.

## Proposed implementation
An implementation of instance generators form Gasse et al. 2019.   The instance generators require a `parameter_generator ` as input, then can be iterated over to generate instances.  A basic view of the python API is provided below: 

```python
import ecole
import numpy as np

class SetcoverParameterGenerator:
    def __init__(self):
        self.rng = np.random.RandomState()

    def __next__(self):
        return {
            'nrows' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'ncols' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'density' : 0.5}  # uniform float in [0.6, 0.8] interval

    def seed(self, seed):
        self.rng.seed(seed)

seed = 0
num_episodes = 100

param_generator = SetcoverParameterGenerator()
param_generator.seed(seed)

instances = ecole.instances.SetcoverGenerator(parameter_generator=param_generator) 
instances.seed(seed)

for i in range(num_episodes):
    instance = next(instances) # returns an ecole.scip.core.Model
    instance.write_problem(f"instance0_{i}.lp") # optional
```

Alternatively, one could iterate over `instances` as shown in the below code example:

```python
import ecole
import numpy as np

class SetcoverParameterGenerator:
    def __init__(self):
        self.rng = np.random.RandomState()

    def __next__(self):
        return {
            'nrows' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'ncols' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'density' : 0.5}  # uniform float in [0.6, 0.8] interval

    def seed(self, seed):
        self.rng.seed(seed)
        
seed = 0
num_episodes = 100
        
param_generator = SetcoverParameterGenerator()
param_generator.seed(seed)

instances = ecole.instances.SetcoverGenerator(parameter_generator=param_generator) 
instances.seed(seed)

episode = 0
for instance in instances:
    if episode == num_episodes:
        break
    instance.write_problem(f"instance1_{episode}.lp") # optional
    episode += 1
```

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+        warnings=False,
+    ):
+        """
+		Generate a Combinatorial Auction instance with specified characteristics, and writes
```
The doc is tab indented, please change to spaces. Missing [short summary](https://numpydoc.readthedocs.io/en/latest/format.html#id8)

AntoinePrv replied:
Check https://github.com/ds4dm/ecole/blob/2d791d0a3b116284e26ae4d3473bf8dde3cd8ada/python/src/ecole/environment.py#L63 for an example

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+			Towards a universal test suite for combinatorial auction algorithms.
+			Proceedings of ACM Conference on Electronic Commerce (EC-00) 66-76.
+		section 4.3., the 'arbitrary' scheme.
+		Parameters
```
Missing empty line before.

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+		integers : logical
+			Should bid's prices be integral ?
+		warnings : logical
+			Should warnings be printed ?
```
Missing empty line after.

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+		section 4.3., the 'arbitrary' scheme.
+		Parameters
+		----------
+		n_items : int
```
Could you put the type annotation as type hints in the signature , like:
```python
def generate_instance(
        self,
        n_items: int = 100,
       ...
)

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+			The resale factor for each bidder, relative to their initial bid's resale value.
+		integers : logical
+			Should bid's prices be integral ?
+		warnings : logical
```
Not a big fan of these printing.

How about
```python
import logging

class CombinatorialAuctionGenerator:
    def __init__(self, parameter_generator=None):
        self.logger = logging.getLogger(self.__class__.__name__)
        ...

    def generate_instance(...):
        ...
        if price < 0:
            # Unconditional. Printing (or not) is configured in the `logging` module
            self.logger.warning("Negatively priced bundle avoided")
            continue
    
```

AntoinePrv commented on python/src/ecole/_setcover_generator.py:
```diff
@@ -0,0 +1,122 @@
+import ecole.scip
+
+
+class SetcoverGenerator:
```
And comments similar to the ones on `CombinatorialAuctionGenerator`

AntoinePrv commented on python/tests/test_instances.py:
```diff
@@ -0,0 +1,40 @@
+    parameter_generator = ParameterGenerator()
+    instances = ecole.instances.SetcoverGenerator(parameter_generator)
+    model = next(instances)
+
```
Could we test the output more here?
Some ideas of predicates to test for could be:
 - Assert problem is minimizing
 - Assert that rhs/matrix coefficients are either 0 or 1
 - Assert variables are binary
 - Assert number of rows is that of the generator
 - Assert number of cols is that of the generator
 - Perhaps check that the problem solves without errors, probably in another test (decorated with `@pytest.mak.slow`)

AntoinePrv commented on python/tests/test_instances.py:
```diff
@@ -0,0 +1,40 @@
+    instances = ecole.instances.CombinatorialAuctionGenerator(parameter_generator)
+    next(instances)
+    model = next(instances)
+
```
Same as previous test.

AntoinePrv commented on python/src/ecole/_combinatorial_auction_generator.py:
```diff
@@ -0,0 +1,229 @@
+import numpy as np
+
+from pyscipopt import Model
```
Could we lazy load PySCIPOpt by moving this import statement inside of `generate_instance`?
PySCIPOpt is optional dependency, but because all modules are imported, this would raise an error.

AntoinePrv reviewed (commented):
I have not yet looked in details at the algorithm but left some minors suggestions.

Additionally, I would personally get rid of the whole `ParameterGenerator` logic and let that up to the user. If they know how to implement a generator, combining an instance generator with it would not be so hard.
Also because when we implement this in C++, It will not have the flexibility it has now and would convolute the bindings code.

jdumouchelle commented:
Thanks for the feedback @AntoinePrv.  I have updated the documentation to be more detailed and implemented the suggestions you have made.  In addition I have removed the logic of `ParameterGenerator` to one that will be compatible with c++ while still providing users with the option to use a generator.  Specifically, users can use `next` if they want to generate an instance with the parameters passed to the constructor of `SetCoverGenerator` or `CombinatorialAuctionGenerator`.  If the user wants to use a generator they can simply directly pass their parameters to `generate_instance`.  Below I have provided an example of both use cases.  

### Using a Parameter generator
```python
import ecole
import numpy as np

class ParameterGenerator:
    def __init__(self):
        self.rng = np.random.RandomState()

    def __next__(self):
        return {
            'nrows' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'ncols' : self.rng.randint(80, 121),  # uniform integer in [80, 120] interval
            'density' : 0.5}  # uniform float in [0.6, 0.8] interval

    def seed(self, seed):
        self.rng.seed(seed)
        
        
param_generator = ParameterGenerator()
param_generator.seed(0)

instances = ecole.instances.SetCoverGenerator() 
instances.seed(0)

num_episodes = 100

for episode in range(num_episodes):
    instance = instances.generate_instance(**next(param_generator)) # returns an ecole.scip.core.Model
    instance.write_problem(f"instance0_{episode}.lp") # optional
```

### Using next or iterating over the class
```python
import ecole
import numpy as np

instances = ecole.instances.SetCoverGenerator(nrows = 500, ncols = 1000, density = 0.05) 
instances.seed(0)

num_episodes = 100

# Example using next
for episode in range(num_episodes):
    instance = next(instances)
    instance.write_problem(f"instance1_{episode}.lp") # optional

# Example by iterating
episode = 0
for instance in instances:
    if episode == num_episodes:
        break
    instance.write_problem(f"instance1_{episode}.lp") # optional
    episode += 1
```

I will add some more detailed testing later today.

AntoinePrv commented:
Posting the result of yesterday discussion
- Let's separate the `generate_instance` into free (stateless) function
  ```python
   def generate_set_cover(param1, param2, ..., random_state) ->  Model
   ```
   That way we have the hard part isolated and we/users can easily built on it to create all kind of generators
- Let's provide the simplest generators for each class of problems, without `ParamGenerators`

AntoinePrv commented:
@jdumouchelle Can we also rename the module `ecole.instances` to `ecole.instance` (singular)? I think it would be more in line with other modules `reward`, `observation`, `environment`.

jdumouchelle commented:
Thanks for posting the notes form the meeting @AntoinePrv .  I have made changes in alignment with what was discussed at the meeting.  

> * Let's separate the `generate_instance` into free (stateless) function
>   ```python
>    def generate_set_cover(param1, param2, ..., random_state) ->  Model
>   ```
>   That way we have the hard part isolated and we/users can easily built on it to create all kind of generators

The changes have been made to make `generate_instance` stateless.

> * Let's provide the simplest generators for each class of problems, without `ParamGenerators`

I have removed the logic for parameter generators and replace it with a class where the parameters are specified in the constructor.  I find a class which a user can iterate over or use `next` over more intuitive than calling a function, but if you strongly prefer we can implement it as a function with `yield` .  

> @jdumouchelle Can we also rename the module `ecole.instances` to `ecole.instance` (singular)? I think it would be more in line with other modules `reward`, `observation`, `environment`.

I agree,  it makes sense to be more in line with the rest of ecole so I have made the change.  

In addition I have added instance generators for independent set and capacitied facility location along with tests for them.

AntoinePrv reviewed (approved):
LGTM. I made a few changes to the structure of the test to remove code duplication and printing to stdout.

I have not looked deeply to the algorithms. I will when moving the generators to C++

gasse commented:
Capacited facility location -> Capacitated facility location

gasse reviewed (changes requested):
Capacited -> capacitated
